atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainerTest.java

```diff
@@ -0,0 +1,325 @@
+import com.palantir.refreshable.Refreshable;
+import com.palantir.refreshable.SettableRefreshable;
+import java.io.Closeable;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import org.awaitility.Awaitility;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ReloadingCloseableContainerTest {
+    private static final ExecutorService executor =
+            PTExecutors.newCachedThreadPool(ReloadingCloseableContainerTest.class.getName());
+
+    private static final int INITIAL_VALUE = 0;
+    private static final int UPDATED_VALUE = 1;
+
+    @Mock
+    private Closeable mockCloseable;
+
+    @Mock
+    private Closeable mockCloseableExtra;
+
+    @Mock
+    private Function<Integer, Closeable> closeableFactory;
+
+    private SettableRefreshable<Integer> settableRefreshable;
+
+    @Before
+    public void setUp() {
+        settableRefreshable = Refreshable.create(INITIAL_VALUE);
+        when(closeableFactory.apply(INITIAL_VALUE)).thenReturn(mockCloseable);
+        when(closeableFactory.apply(UPDATED_VALUE)).thenReturn(mockCloseableExtra);
+    }
+
+    @Test
+    public void closeBlocksDuringRunWithInvocation() throws BrokenBarrierException, InterruptedException {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+        CyclicBarrier runWithFutureRunsFirstBarrier = new CyclicBarrier(2);
+        CyclicBarrier manuallyFinishRunWith = new CyclicBarrier(2);
+
+        executor.execute(() -> container.executeWithResource(closeable -> {
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            awaitOrFail(manuallyFinishRunWith);
+            verifyNoInteractions(closeable);
+        }));
+
+        // This test could pass spuriously, if close is just not called.
+        // The likelihood if this is minimised by
+        // 1. waiting on a barrier for the executeWithResource and,
+        // 2. Requiring a whole second of close not completing.
+        Future<?> closeFuture = executor.submit(() -> {
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            container.close();
+        });
+        Awaitility.await().during(Duration.ofSeconds(1)).until(() -> !closeFuture.isDone());
+        verifyNoInteractions(mockCloseable);
+
+        manuallyFinishRunWith.await();
+        Awaitility.await().atMost(Duration.ofSeconds(1)).until(closeFuture::isDone);
+    }
+
+    @Test
+    public void refreshCloseBlockedDuringRunWithInvocation()
```

Comment:
Is this what this test is testing?
```java
        /*
         * --- runWithResource start
         *  |
         *  |
         * --- refresh value
         *  |
         *  |
         * --- runWithResource finish with original resource
         */
```

Reply:
Wait, no, I think it's just confirming that the original is closed at the end, but the new is available straight after refresh (although it doesn't test that second part)